rfcs/0000-unified-configuration.ftd

```diff
@@ -0,0 +1,188 @@
+-- import: fastn.com/rfcs/lib
+-- import: fastn-community.github.io/bling/note
+
+-- lib.rfc: RFC: Unified Configuration
+id: unified-config
+status: proposal
+
+TODO: Write a brief summary of the RFC here.
+
+
+-- lib.motivation:
+
+;; TODO: add links to the following paragraph
+Currently, `fastn` supports configuration through environment variables, and in
+some way, through the `FASTN.ftd` file. Environment variables are good to
+change the behaviour of the cli or to pass secret information but it's not
+ideal for defining complex configuration for a framework like `fastn` that
+provides various configurable features like processors, oauth authentication
+and, apps.
+
+;; TODO: I can do better than this
+This RFC proposes a way of dealing with configuration that is centralised and
+leveraged the existing ability of the `fastn` cli to read environment
+variables.
+
+-- end: lib.motivation
+
+
+-- lib.detailed-design:
+
+-- ds.h2: `fastn/config` - A virtual module to assist configuration definition.
+
+Add a virtual module (`fastn/config`) similar to existing [assets](TODO: add
+link) module.
+
+;; TODO add footer section for explanors like foreign variables and virtual module
+- This module contains all special configuration types defined by the fastn
+  framework. This module also provides safe access to environment variables
+  through foreign variables.
+
+- `FASTN.ftd` file will be central to the whole configuration system. Users can
+  modify the behaviour of a dependency by passing variables when they're added.
+  Example:
+
+-- ds.code:
+lang: ftd
+
+;; FASTN.ftd of test-todos
+\-- import: fastn
+
+\-- fastn.package: test-todos
+
+\;; fastn.app works in similar way
+\-- fastn.dependency: amitu.com/todos
+max-todo-per-page: 30
+ignore-value: test \;; amitu.com/todos doesn't have this variable so we ignore this
```

Comment:
Why are we having this? I think we should throw error in this case.

Reply:
Our general philosophy is to be as strict as possible. When we become lax we have to handle a lot of cases, do a lot of guess work. With strictness comes simplicity. User may not like errors, but they def hate complexity (and the bugs and wastage of time comes due to them).